modules/stochastic_tools/include/reporters/ActiveLearningReporterBase.h

```diff
@@ -0,0 +1,152 @@
+   *      or not.
+   */
+  virtual ReporterName declareStochasticReporterClone(const Sampler & sampler,
+                                                      const ReporterData & from_data,
+                                                      const ReporterName & from_reporter,
+                                                      std::string prefix = "") override;
+
+  /**
+   * Get a const reference to the sampler from the parameters. This is preferred
+   * over having _sampler being a protected member since we don't want derived classes
+   * changing the state of the sampler during the loop in execute.
+   */
+  const Sampler & sampler() const { return _sampler; }
+
+  /**
+   * This routine is called during the sampler loop in execute() and is meant to fill
+   * in the "need_sample" reporter value and modify the data declared by the transfer.
+   *
+   * @param row The row of data from the sampler
+   * @param local_ind The local index of the sampler row
+   * @param global_ind The global index of the sampler row
+   * @param val Reference to the value associated with the row of data.
+   * @return Filled in value of "need_sample", meant to determine if a sample needs
+   *         to be run by a multiapp or other means.
+   */
+  virtual bool needSample(const std::vector<Real> & /*row*/,
+                          dof_id_type /*local_ind*/,
+                          dof_id_type /*global_ind*/,
+                          T & /*val*/)
+  {
+    return true;
+  }
+
+private:
+  /// Sampler given in the parameters, must match the one used to declare the
+  /// transferred values.
+  Sampler & _sampler;
+  /// Reporter value determining whether we need to evaluate the sample through
+  /// a multiapp or other means.
+  std::vector<bool> & _need_sample;
+  /// Reporter value declared with the transfer
+  std::vector<T> * _data = nullptr;
+};
+
+template <typename T>
+InputParameters
+ActiveLearningReporterTempl<T>::validParams()
+{
+  InputParameters params = StochasticReporter::validParams();
+  params.addRequiredParam<SamplerName>("sampler", "The sampler used to produce data.");
+  return params;
+}
+
+template <typename T>
+ActiveLearningReporterTempl<T>::ActiveLearningReporterTempl(const InputParameters & parameters)
+  : StochasticReporter(parameters),
+    _sampler(this->template getSampler<Sampler>("sampler")),
+    _need_sample(this->template declareStochasticReporter<bool>("need_sample", _sampler))
+{
+}
+
+template <typename T>
+void
+ActiveLearningReporterTempl<T>::execute()
+{
+  // Dummy value in case _data has not been declared yet
+  T dummy;
+  // Loop over samples to determine if sample is needed. Replace value in _data
+  // (typically only if a sample is not needed). We insert a dummy value in case
+  // _data has not been declared.
+  for (const auto & i : make_range(_sampler.getNumberOfLocalRows()))
+    _need_sample[i] = needSample(_sampler.getNextLocalRow(),
```

Comment:
I get it now. The reason you have two dof_id_type in need sample is two support distributed sampling, correct?

Reply:
Correct.